lib/game_runner.rb

```diff
@@ -1,37 +1,43 @@
 require_relative "grid"
 require_relative "printer"
+require_relative "player_input"
+require_relative "grid_printer"
 
 class GameRunner
-  attr_reader :stdout, :stdin, :grid, :printer
+  attr_reader :stdout, :stdin, :grid, :printer, :player_input, :grid_printer
 
   def initialize(stdout, stdin, grid = Grid.new)
     @stdout = stdout
     @stdin = stdin
     @grid = grid
-  end
-
-  def printer
-    @printer ||= Printer.new(stdout, grid)
+    @printer ||= Printer.new(stdout)
```

Comment:
Since you are setting the `@printer` inside initialize, there is never going to be a situation where it might already exist. So using the `||=` isn't a good approach. Was this copied from somewhere else?

Relatedly, I tend to prefer the instantiation of classes to happen in their own method. That way if something goes wrong in the instantiation, it is easier to debug. If you moved these to methods, then you would use `||=` and you wouldn't have `attr_reader`s for them.

Reply:
Thank you! Yes. I was following something from the dependency injection article you had sent me in an older feature, and I don't think I fully understood the reasoning behind it. It makes sense now!